processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -157,6 +183,101 @@ func (mtp *metricsTransformProcessor) ProcessMetrics(_ context.Context, md pdata
+		}
+		if metric.MetricDescriptor.Unit != firstMetricDescriptor.Unit {
+			return fmt.Errorf("metrics cannot be combined as they have different units: %v (%v) and %v (%v)", firstMetricDescriptor.Name, firstMetricDescriptor.Unit, metric.MetricDescriptor.Name, metric.MetricDescriptor.Unit)
+		}
+
+		if len(metric.MetricDescriptor.LabelKeys) != len(firstMetricLabelKeys) {
+			return fmt.Errorf("metrics cannot be combined as they have different labels: %v (%v) and %v (%v)", firstMetricDescriptor.Name, firstMetricDescriptor.LabelKeys, metric.MetricDescriptor.Name, metric.MetricDescriptor.LabelKeys)
+		}
+
+		for _, label := range metric.MetricDescriptor.LabelKeys {
+			if _, ok := firstMetricLabelKeys[label.Key]; !ok {
+				return fmt.Errorf("metrics cannot be combined as they have different labels: %v (%v) and %v (%v)", firstMetricDescriptor.Name, firstMetricDescriptor.LabelKeys, metric.MetricDescriptor.Name, metric.MetricDescriptor.LabelKeys)
+			}
+		}
+	}
+
+	return nil
+}
+
+// combine combines the metrics based on the supplied filter.
+func (mtp *metricsTransformProcessor) combine(matchedMetrics []*match, transform internalTransform) *metricspb.Metric {
+	// create combined metric with relevant name & descriptor
+	combinedMetric := &metricspb.Metric{}
+	combinedMetric.MetricDescriptor = proto.Clone(matchedMetrics[0].metric.MetricDescriptor).(*metricspb.MetricDescriptor)
+	combinedMetric.MetricDescriptor.Name = transform.NewName
+	combinedMetric.MetricDescriptor.Description = ""
+
+	// append label keys based on the transform filter's named capturing groups
+	subexprNames := transform.MetricIncludeFilter.getSubexpNames()
+	for i := 1; i < len(subexprNames); i++ {
+		// if the subexpression is not named, use regexp notation, e.g. $1
+		name := subexprNames[i]
+		if name == "" {
+			name = "$" + strconv.Itoa(i)
+		}
+
+		combinedMetric.MetricDescriptor.LabelKeys = append(combinedMetric.MetricDescriptor.LabelKeys, &metricspb.LabelKey{Key: name})
+	}
+
+	// combine timeseries from all metrics, using the specified AggregationType if data points need to be merged
+	var allTimeseries []*metricspb.TimeSeries
+	for _, match := range matchedMetrics {
+		for _, ts := range match.metric.Timeseries {
+			// append label values based on regex submatches
+			for i := 1; i < len(match.submatches)/2; i++ {
+				submatch := match.metric.MetricDescriptor.Name[match.submatches[2*i]:match.submatches[2*i+1]]
+				ts.LabelValues = append(ts.LabelValues, &metricspb.LabelValue{Value: submatch, HasValue: (submatch != "")})
+			}
+
+			allTimeseries = append(allTimeseries, match.metric.Timeseries...)
+		}
+	}
+
+	groupedTimeseries := mtp.groupTimeseries(allTimeseries, len(combinedMetric.MetricDescriptor.LabelKeys))
+	aggregatedTimeseries := mtp.mergeTimeseries(groupedTimeseries, transform.AggregationType, combinedMetric.MetricDescriptor.Type)
```

Comment:
Do you validate that `transform.AggregationType` is set correctly? Similar to Action being one of 3 possible "enum" values.

Reply:
Also, should the function be named `combineTimeseries`, or action "merge" for consistency?

Reply:
> Do you validate that `transform.AggregationType` is set correctly? Similar to Action being one of 3 possible "enum" values.

The way config is validated in this component is a little strange at the moment and there is validation missing for quite a few important cases. For now, I've added the obvious missing validation: on `AggregationType` & `OperationAction`.

This component probably needs a future PR to separate config into different types for different actions.

> Also, should the function be named `combineTimeseries`, or action "merge" for consistency?

I think this is okay. One aspect of the "combine" action is merging timeseries. Merging timeseries is also relevant for the aggregating labels & aggregating label values actions.